File: data-structures/linear/stack/Stack.cpp
#include <iostream>
using namespace std;

class Node {
public:
  int data;
  Node *next;

  Node(int value) {
    data = value;
    next = nullptr;
  }
};

class Stack {
  Node *head;
  int length;

public:
  Stack() {
    head = nullptr;
    length = 0;
  }

  void push(int value) {
    Node *newNode = new Node(value);

    if (head == nullptr) {
      head = newNode;
      length++;
      return;
    }

    newNode->next = head;
    head = newNode;
    length++;
  }

  void pop() {
    if (head == nullptr) {
      cout << "Stack is empty!\n";
      return;
    }

    Node *temp = head;
    
    head = head->next;
    temp->next = nullptr;
    delete temp;
    length--;
  }

  int top() { return head->data; }

  bool empty() { return length == 0; }

  int size() { return length; }
};

File: data-structures/linear/stack/Stack.ts
//@ts-expect-error
import { process } from "node:process";
import type { Node } from "../node/Node";

class Stack {
	private head: Node | null;
	private size: number;

	constructor(head: Node | null = null) {
		this.head = head;
		this.size = 0;
	}

	public getHead(): Node | null {
		return this.head;
	}

	public getSize(): number {
		return this.size;
	}

	public push(node: Node): void {
		node.setNextNode(this.head);
		this.head = node;
	}

	public pop(): Node {
		if (this.head == null) {
			throw Error("Stack UnderFlow Exception");
		}

		const poppedNode: Node = this.head;
		this.head = this.head.getNextNode();
		return poppedNode;
	}

	public top(): Node {
		if (this.head == null) {
			throw Error("Stack UnderFlow Exception");
		}

		return this.head;
	}

	public display(): void {
		let n: Node | null = this.head;
		for (; n?.getNextNode() != null; n = n.getNextNode())
			process.std.write(`${n.getValue()} -> `);
		process.std.write();
	}
}

export { Stack };
